extendr-api/src/robj_ndarray.rs

```diff
@@ -55,6 +57,34 @@ make_array_view_2!(f64, "Not a floating point matrix.");
 //         r!(mx)
 //     }
 // }
+impl<A, S, D> From<ArrayBase<S, D>> for Robj
+where
+    S: Data<Elem = A>,
+    A: Copy + ToVectorValue,
+    D: Dimension,
+{
+    fn from(arr: ArrayBase<S, D>) -> Self {
+        let dims: Vec<i32> = arr.shape().iter().map(|x| *x as i32).collect();
+        let mut ret = arr
+            .as_slice()
+            .ok_or("Could not convert")
+            .unwrap()
+            .iter()
+            .copied()
+            .collect_robj();
+
+        if dims.len() > 1 {
+            ret = ret.set_attrib(dim_symbol(), dims.clone()).unwrap();
+            if dims.len() == 2 {
+                ret = ret.set_attrib(class_symbol(), ["matrix", "array"]).unwrap()
+            } else {
+                ret = ret.set_attrib(class_symbol(), ["array"]).unwrap()
+            }
+        }
```

Comment:
`matrix` and `array` are not S3 classes. I think you can just set `.Dim` attribute unconditionally (I mean, even when `dim.len() == 1`).

``` r
f <- function(dim) {
  x <- array(1:6, dim = dim)
  
  print(x)
  
  # check class
  print(is(x))
  
  # check the serialized form (confirm there's no class)
  dput(x)
  
  # check the internal representation
  .Internal(inspect(x))
}

# vec
f(6L)
#> [1] 1 2 3 4 5 6
#> [1] "array"     "structure" "vector"   
#> structure(1:6, .Dim = 6L)
#> @55a9750033f8 13 INTSXP g0c3 [REF(1),ATT] (len=6, tl=0) 1,2,3,4,5,...
#> ATTRIB:
#>   @55a975ddc598 02 LISTSXP g0c0 [REF(1)] 
#>     TAG: @55a974082ac0 01 SYMSXP g0c0 [MARK,REF(2694),LCK,gp=0x4000] "dim" (has value)
#>     @55a975f1e9b0 13 INTSXP g0c1 [REF(65535)] (len=1, tl=0) 6

# matrix
f(c(2L, 3L))
#>      [,1] [,2] [,3]
#> [1,]    1    3    5
#> [2,]    2    4    6
#> [1] "matrix"    "array"     "structure" "vector"   
#> structure(1:6, .Dim = 2:3)
#> @55a977c09e98 13 INTSXP g0c3 [REF(1),ATT] (len=6, tl=0) 1,2,3,4,5,...
#> ATTRIB:
#>   @55a9766f52a8 02 LISTSXP g0c0 [REF(1)] 
#>     TAG: @55a974082ac0 01 SYMSXP g0c0 [MARK,REF(2732),LCK,gp=0x4000] "dim" (has value)
#>     @55a9775b4ef0 13 INTSXP g0c1 [REF(65535)] (len=2, tl=0) 2,3

# array
f(c(2L, 3L, 1L))
#> , , 1
#> 
#>      [,1] [,2] [,3]
#> [1,]    1    3    5
#> [2,]    2    4    6
#> 
#> [1] "array"     "structure" "vector"   
#> structure(1:6, .Dim = c(2L, 3L, 1L))
#> @55a977c24ce8 13 INTSXP g0c3 [REF(1),ATT] (len=6, tl=0) 1,2,3,4,5,...
#> ATTRIB:
#>   @55a97678e768 02 LISTSXP g0c0 [REF(1)] 
#>     TAG: @55a974082ac0 01 SYMSXP g0c0 [MARK,REF(2735),LCK,gp=0x4000] "dim" (has value)
#>     @55a977c0f268 13 INTSXP g0c2 [REF(65535)] (len=3, tl=0) 2,3,1
```

<sup>Created on 2021-09-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Reply:
I'm not entirely sure what this means. Are you saying that they don't have a class symbol set ever?

Reply:
Yes.

Reply:
Yes, correct. Compare these two. First does not have a class symbol, second does.

``` r
x <- array(1:6, dim = 2)
dput(x)
#> structure(1:2, .Dim = 2L)

x <- grid::unit(4, "in")
dput(x)
#> structure(4, unit = 2L, class = c("simpleUnit", "unit", "unit_v2"
#> ))
```

<sup>Created on 2021-08-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Reply:
> matrix and array are not S3 classes. I think you can just set .Dim attribute unconditionally (I mean, even when dim.len() == 1).

This doesn't appear to be true. If I set dim unconditionally the comparison (implemented in R's API) fails.

Reply:
I guess that's because you are using `c()` for testing 1D array. I'm suggesting returning a 1D array instead of a vector.

Reply:
Right, but I think that R users will expect 1D outputs from any function to be as a vector rather than a 1D array. 1D arrays are basically unheard of to my knowledge.

Reply:
Or would you rather that this conversion always returns an array, for consistency?

Reply:
Yes, for the sake of consistency. But, I'm not fully sure what's the right thing to do here. I'd like to wait for others' comments.

> 1D arrays are basically unheard of to my knowledge.

I think that's because 1D arrays can be handled just the same as usual vectors transparently. 1D array is probably not useful, but should be harmless.

Reply:
You could be right, they're probably harmless.

Reply:
Yeah look I'm tending towards agreeing now. It means a simpler implementation. And if `extendr` users want to return an R vector they can just return [`arr.to_vec()`](https://docs.rs/ndarray/0.15.3/ndarray/struct.ArrayBase.html#method.to_vec) to do this explicitly (I assume we already have a conversion between R and Rust vectors).